main.go

```diff
@@ -80,212 +76,235 @@ func main() {
-					// Doesn't instance exist in update?
-					if _, ok := update.ServiceInstances[k]; !ok {
-						toRemove = append(toRemove, gce.NetworkEndpoint{
-							Instance: normalizeInstanceName(instance.Host),
-							IP:       instance.Address,
-							Port:     instance.Port,
-						})
-						delete(instances, k)
-					}
+				isRunning = true
+			case registry.DELETED:
+				if !isRunning {
+					glog.Infof("Ignored DELETED update of service with tag %q since it's not running", tagName)
+					continue
 				}
-
-				// finding instances to add
-				for k, instance := range update.ServiceInstances {
-					// Doesn't instance exist in running instances?
-					if _, ok := instances[k]; !ok {
-						toAdd = append(toAdd, gce.NetworkEndpoint{
-							Instance: normalizeInstanceName(instance.Host),
-							IP:       instance.Address,
-							Port:     instance.Port,
-						})
-						instances[k] = instance
-					}
+				if err := handleServiceDeletedUpdate(client, tagConfig, lock, instances); err != nil {
+					glog.Errorf(errorTemplate, update.UpdateType, tagName, err)
+					continue
 				}
-
-				//do we have instances to remove from the NEG?
-				if len(toRemove) > 0 {
-					if err := client.DetachEndpointsFromGroup(toRemove, makeName("neg", serviceGroupName)); err != nil {
-						glog.Errorf("Failed detaching endpoints from %s network endpoint group: %v", makeName("neg", serviceGroupName), err)
-					}
+				isRunning = false
+			case registry.CHANGED:
+				if !isRunning {
+					glog.Infof("Ignored CHANGED update of service with tag %q since it's not running", tagName)
+					continue
 				}
-
-				// do we have new instances to add to the NEG?
-				if len(toAdd) > 0 {
-					if err := client.AttachEndpointsToGroup(toAdd, makeName("neg", serviceGroupName)); err != nil {
-						glog.Errorf("Failed adding endpoints to %s network endpoint group: %v", makeName("neg", serviceGroupName), err)
-					}
+				if err := handleServiceChangedUpdate(client, tagConfig, lock, instances, update); err != nil {
+					glog.Errorf(errorTemplate, update.UpdateType, tagName, err)
 				}
-
-				lock.Unlock()
-			default:
-				continue
 			}
 		case <-done:
-			glog.Warningf("Received termination signal for service with tag %s", tag)
+			glog.Warningf("Received termination signal for service with tag %q", tagName)
 			wg.Done()
 			return
 		}
 	}
 }
 
-// NOTE: Maximum length of resource name is 63 and name must start and end with letter or digit
-func makeName(prefix string, name string) string {
-	n := strings.Join([]string{prefix, name}, "-")
-	if len(n) > maximumNameLength {
-		n = strings.TrimLeft(n[:maximumNameLength], "-")
+func handleServiceNewUpdate(
+	urlMapName string,
+	client cloud.Cloud,
+	tagConfig *tagConfiguration,
+	lock *sync.RWMutex,
+) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	tagName := tagConfig.Name
+
+	if tagConfig.Mode == REUSE_MODE {
+		glog.Infof("Service with tag %q will use externally managed resources", tagName)
+	} else {
```

Comment:
All your _main_ logic is in the `else` statement. It is more common to have the main code path in the main thread.

Reply:
I would say there are no main logic since these are just two approches 🤔